---
title: "Ball Horticultural Coding Challenge"
output: html_notebook
---
Download/Import dependencies if needed
```{r}
install.packages("gplots")
library(gplots)
```

First of all, I will import the data file and do a little bit of EDA to dive into the data and see what I'm working with. 
I will assume here that 3 replicates were taken, baseline expression for each gene was measured, the bud was then given a treatment, and then the change in expression of each gene was measured. 
```{r}
exprs.data <- read.delim("test_gene_expression_matrix.txt", header = TRUE, sep = "\t")
print(head(exprs.data))

#confirm that all genes are present only once
#looks like there are ~29k instances of genes with no repeats
cat("Number of unique genes:", length(unique(exprs.data$Name)))
cat("\nNumber of rows:", length(exprs.data$Name), "\n")



#find the most highly expressed gene for each of the samples tested, these seem to be Flower_024636 and Flower_010909-RA for the untreated group, and Flower_011193-RA for the treated group
#original replicates:
cat("\n","Maximum expression of original replicates:","\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Bud_Rep1)]), max(exprs.data$Bud_Rep1), "\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Bud_Rep2)]), max(exprs.data$Bud_Rep2), "\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Bud_Rep3)]), max(exprs.data$Bud_Rep3), "\n")

#treated replicates:
cat("\n","Maximum expression of treated replicates:","\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Day2_Rep1)]), max(exprs.data$Day2_Rep1), "\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Day2_Rep2)]), max(exprs.data$Day2_Rep2), "\n")
cat(as.character(exprs.data$Name[which.max(exprs.data$Day2_Rep3)]), max(exprs.data$Day2_Rep3), "\n")



#find the average expression of all genes for each replicate and standard deviation
#I don't think that this information would be particularly useful, but I was curious anyway. It looks like the average expression tends to be around 34
#original replicates:
cat("\n","Mean expression of original replicates:","\n")
cat(mean(exprs.data$Bud_Rep1), sd(exprs.data$Bud_Rep1), "\n")
cat(mean(exprs.data$Bud_Rep2), sd(exprs.data$Bud_Rep2), "\n")
cat(mean(exprs.data$Bud_Rep3), sd(exprs.data$Bud_Rep3), "\n")

#treated replicates:
cat("\n","Mean expression of treated replicates:","\n")
cat(mean(exprs.data$Day2_Rep1), sd(exprs.data$Bud_Rep1), "\n")
cat(mean(exprs.data$Day2_Rep2), sd(exprs.data$Bud_Rep2), "\n")
cat(mean(exprs.data$Day2_Rep3), sd(exprs.data$Bud_Rep3), "\n")



#Finally, I thought it may be nice to see how many genes are expressed in each sample. In other words, how many genes are nonzero.
#original
cat("\n","Total number of genes expressed in original replicates, proportion expressed","\n")
cat(length(which(exprs.data$Bud_Rep1 > 0)), length(which(exprs.data$Bud_Rep1 > 0))/29333,"\n")
cat(length(which(exprs.data$Bud_Rep2 > 0)), length(which(exprs.data$Bud_Rep2 > 0))/29333,"\n")
cat(length(which(exprs.data$Bud_Rep3 > 0)), length(which(exprs.data$Bud_Rep3 > 0))/29333,"\n")

#treated
cat("\n","Total number of genes expressed in original replicates, proportion expressed","\n")
cat(length(which(exprs.data$Day2_Rep1 > 0)), length(which(exprs.data$Day2_Rep1 > 0))/29333,"\n")
cat(length(which(exprs.data$Day2_Rep2 > 0)), length(which(exprs.data$Day2_Rep2 > 0))/29333,"\n")
cat(length(which(exprs.data$Day2_Rep3 > 0)), length(which(exprs.data$Day2_Rep3 > 0))/29333,"\n")
```

In a data analysis project I think it is important to establish the question to be answered, and define a minimum viable product that should be able to be presented and/or submitted.

In this case I defined the question to be: Does the treatment result in any observable differences in gene expression over the 3 replicates between the original and Day2 treatments?

My MVP would be (as described in the instructions): How many genes in each replicate are differentially expressed after the treatment?

Additionally, I would plan to include some sort of heatmap in order to clearly communicate the change in gene expression to all parties involved. I would use the top 10 (20?) most differentially expressed genes that we found using this treatment based on 3 replicates in order to condense the data into something that is easily visualized.



For the actual analysis, I will first create additional columns to represent the change in expression in each gene over the 3 replicates, and then find the number of differentially expressed genes. I decided to start off with looking at genes in which the difference in expression was greater than 0.
```{r}
exprs.data$Delta_Rep1 <- exprs.data$Day2_Rep1-exprs.data$Bud_Rep1
exprs.data$Delta_Rep2 <- exprs.data$Day2_Rep2-exprs.data$Bud_Rep2
exprs.data$Delta_Rep3 <- exprs.data$Day2_Rep3-exprs.data$Bud_Rep3

#also another column for the average change across 3 replicates
for (i in 1:nrow(exprs.data)){
  exprs.data$Delta_Ave[i] = sum(exprs.data$Day2_Rep1[i], exprs.data$Day2_Rep2[i], exprs.data$Day2_Rep3[i])/3 - sum(exprs.data$Bud_Rep1[i], exprs.data$Bud_Rep2[i], exprs.data$Bud_Rep3[i])/3
}

#How many genes per replicate show ANY change in expression between the starting and day2 values?
#rep1
print("Number of genes differentially expressed in Rep1:")
length(which(abs(exprs.data$Delta_Rep1) > 0))

#rep2
print("Number of genes differentially expressed in Rep2:")
length(which(abs(exprs.data$Delta_Rep2) > 0))

#rep3
print("Number of genes differentially expressed in Rep3:")
length(which(abs(exprs.data$Delta_Rep3) > 0))
```

This analysis shows that ~20,000 genes are differentially expressed. This is a lot to consider, and I would assume that part of this difference in expression should be written off as natural biological variance and/or background noise.

Finding the "top 10" upregulated and downregulated genes based on overall variance for each replicate:
```{r}
#Return the top and bottom 10 genes from each replicate:
#rep1
cat("Rep 1 Top 10 upregulated genes:","\n")
#indicies of top 10 genes
ind.rep1.up <- head(order(-exprs.data$Delta_Rep1), 10)
up.rep1 <- exprs.data$Name[ind.rep1.up]
print(as.character(up.rep1))
up.rep1.val <- exprs.data$Delta_Rep1[ind.rep1.up]
print(up.rep1.val)

cat("Rep 1 Top 10 downregulated genes:","\n")
ind.rep1.down <- head(order(exprs.data$Delta_Rep1), 10)
down.rep1 <- exprs.data$Name[ind.rep1.down]
print(as.character(down.rep1))
down.rep1.val <- exprs.data$Delta_Rep1[ind.rep1.down]
print(down.rep1.val)


#rep2
cat("Rep 2 Top 10 upregulated genes:","\n")
ind.rep2.up <- head(order(-exprs.data$Delta_Rep2), 10)
up.rep2 <- exprs.data$Name[ind.rep2.up]
print(as.character(up.rep2))
up.rep2.val <- exprs.data$Delta_Rep2[ind.rep2.up]
print(up.rep2.val)

cat("Rep 2 Top 10 downregulated genes:","\n")
ind.rep2.down <- head(order(exprs.data$Delta_Rep2), 10)
down.rep2 <- exprs.data$Name[ind.rep2.down]
print(as.character(down.rep2))
down.rep2.val <- exprs.data$Delta_Rep2[ind.rep2.down]
print(down.rep2.val)


#rep3
cat("Rep 3 Top 10 upregulated genes:","\n")
ind.rep3.up <- head(order(-exprs.data$Delta_Rep3), 10)
up.rep3 <- exprs.data$Name[ind.rep3.up]
print(as.character(up.rep3))
up.rep3.val <- exprs.data$Delta_Rep2[ind.rep3.up]
print(up.rep3.val)

cat("Rep 3 Top 10 downregulated genes:","\n")
ind.rep3.down <- head(order(exprs.data$Delta_Rep3), 10)
down.rep3 <- exprs.data$Name[ind.rep3.down]
print(as.character(down.rep3))
down.rep3.val <- exprs.data$Delta_Rep2[ind.rep3.down]
print(down.rep3.val)

#Overall
cat("Overall Top 10 upregulated genes:","\n")
ind.ave.up <- head(order(-exprs.data$Delta_Ave), 10)
up.ave <- exprs.data$Name[ind.ave.up]
print(as.character(up.ave))
up.ave.val <- exprs.data$Delta_Ave[ind.ave.up]
print(up.ave.val)

cat("Overall Top 10 downregulated genes:","\n")
ind.ave.down <- head(order(exprs.data$Delta_Ave), 10)
down.ave <- exprs.data$Name[ind.ave.down]
print(as.character(down.ave))
down.ave.val <- exprs.data$Delta_Ave[ind.ave.down]
print(down.ave.val)
```


Heatmaps
I tried applying the scale parameter here in order to place all of the expression values between -1 and 1, but it resulted in little color variation.
```{r}
#generate red to green color scheme
redgreen <- function(n) {
c(hsv(h=2/6, v=seq(1,0,length=n/2) ), hsv(h=0/6, v=seq(0,1,length=n/2)) )
}

#rep1
top20.rep1 <- c(ind.rep1.up, ind.rep1.down)
top20.rep1.heatmap <- data.matrix(exprs.data[top20.rep1, c("Bud_Rep1", "Day2_Rep1")])

heatmap.2(top20.rep1.heatmap, main = "Rep1 Top 20 Genes", labRow = exprs.data$Name[top20.rep1], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#rep2
top20.rep2 <- c(ind.rep2.up, ind.rep2.down)
top20.rep2.heatmap <- data.matrix(exprs.data[top20.rep2, c("Bud_Rep2", "Day2_Rep2")])

heatmap.2(top20.rep2.heatmap, main = "Rep2 Top 20 Genes", labRow = exprs.data$Name[top20.rep2], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#rep3
top20.rep3 <- c(ind.rep3.up, ind.rep3.down)
top20.rep3.heatmap <- data.matrix(exprs.data[top20.rep3, c("Bud_Rep3", "Day2_Rep3")])

heatmap.2(top20.rep3.heatmap, main = "Rep3 Top 20 Genes", labRow = exprs.data$Name[top20.rep3], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#Overall
top20.ave <- c(ind.ave.up, ind.ave.down)
top20.ave.heatmap <- data.matrix(exprs.data[top20.ave, c("Bud_Rep1", "Bud_Rep2", "Bud_Rep3", "Day2_Rep1", "Day2_Rep2", "Day2_Rep3")])

heatmap.2(top20.ave.heatmap, main = "Overall Top 20 Genes", labRow = exprs.data$Name[top20.ave], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")
```


I experimented with taking the log of the expression values rather than raw values, to see if it created a clearer heatmap. I think that these are somewhat clearer to read than the previous heatmaps, and result in a more easy to interpret visualization.
```{r}
#overwrites previous variables!!!!
#rep1
top20.rep1 <- c(ind.rep1.up, ind.rep1.down)
top20.rep1.heatmap <- data.matrix(log(exprs.data[top20.rep1, c("Bud_Rep1", "Day2_Rep1")]))

heatmap.2(top20.rep1.heatmap, main = "Rep1 Top 20 Genes", labRow = exprs.data$Name[top20.rep1], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#rep2
top20.rep2 <- c(ind.rep2.up, ind.rep2.down)
top20.rep2.heatmap <- data.matrix(log(exprs.data[top20.rep2, c("Bud_Rep2", "Day2_Rep2")]))

heatmap.2(top20.rep2.heatmap, main = "Rep2 Top 20 Genes", labRow = exprs.data$Name[top20.rep2], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#rep3
top20.rep3 <- c(ind.rep3.up, ind.rep3.down)
top20.rep3.heatmap <- data.matrix(log(exprs.data[top20.rep3, c("Bud_Rep3", "Day2_Rep3")]))

heatmap.2(top20.rep3.heatmap, main = "Rep3 Top 20 Genes", labRow = exprs.data$Name[top20.rep3], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")

#Overall
top20.ave <- c(ind.ave.up, ind.ave.down)
top20.ave.heatmap <- data.matrix(log(exprs.data[top20.ave, c("Bud_Rep1", "Bud_Rep2", "Bud_Rep3", "Day2_Rep1", "Day2_Rep2", "Day2_Rep3")]))

heatmap.2(top20.ave.heatmap, main = "Overall Top 20 Genes", labRow = exprs.data$Name[top20.ave], dendrogram = "none", margins = c(7,7), cexCol = 1, cexRow = 0.6, Rowv = FALSE, Colv = FALSE, col = redgreen(100), trace = "none")
```


Next, I wanted to allow the heatmap to create a dendrogram of genes, therefore clustering genes together that show a similar expression pattern between all samples. The thought behind this would be that it may be beneficial to look into highly variable genes that consistently show similar expression patterns between the Bud and Day2 treatments.

I decided to do this using a total of 50 genes (25 from both up- and down-regulated expression). I also decided to center and scale the expression values across the 6 samples in order to highlight the expression differences rather than the raw expression levels (as in the earlier heatmaps). This process could be repeated using a larger number of "top" genes, in order to identify genes that display a smaller level of variance, however, it will be even harder to visualize.
```{r}
#this also overwrites previously defined variables!!

#change heatmap to include top 25 instead of 10 and draw dendogram
cat("Overall Top 25 upregulated genes:","\n")
ind.ave.up <- head(order(-exprs.data$Delta_Ave), 25)
up.ave <- exprs.data$Name[ind.ave.up]
print(as.character(up.ave))
up.ave.val <- exprs.data$Delta_Ave[ind.ave.up]
print(up.ave.val)

cat("Overall Top 25 downregulated genes:","\n")
ind.ave.down <- head(order(exprs.data$Delta_Ave), 25)
down.ave <- exprs.data$Name[ind.ave.down]
print(as.character(down.ave))
down.ave.val <- exprs.data$Delta_Ave[ind.ave.down]
print(down.ave.val)


top50.ave <- c(ind.ave.up, ind.ave.down)
top50.ave.heatmap <- data.matrix(log(exprs.data[top50.ave, c("Bud_Rep1", "Bud_Rep2", "Bud_Rep3", "Day2_Rep1", "Day2_Rep2", "Day2_Rep3")]))

png("high_res_clust.png", width = 7, height = 5, units = 'in', res = 600)
heatmap.2(top50.ave.heatmap, main = "Overall Top 50 Genes, Clustered", labRow = exprs.data$Name[top50.ave], dendrogram = "row", margins = c(7,7), scale = "row", col = redgreen(100), cexCol = 1, cexRow = 0.3, trace = "none")
```

If available, I would utilize an annotation package such as those that are available from Bioconductor (https://bioconductor.org/packages/release/data/annotation/) in order to run enrichment analysis to see if any of the clustered genes are associated, or what sort of association they have.

In conclusion, I would look further into the following genes:

Flower_031266-RA
Flower_025536-RA
Flower_050923-RA
Flower_010327-RA
Flower_024636-RA
Flower_031552-RA
Flower_045575-RA
Flower_030436-RA

Based on the clustering analysis, I would also want to look at these genes:

Flower_003029-RA
Flower_008444-RA
Flower_033580-RA
